### ISSUE TITLE
genhtml Failing with error cannot create directory build/test

### ISSUE THREAD

**Issue by rachit16** [linux-test-project/lcov#183]
Getting below error while running genhtml command :

sudo genhtml --rc lcov_branch_coverage=1 -s -k --legend -p /home/root/workspace/test-build test.info -o /root/reports/test/html-test --ignore-errors source

23:42:12  01-Jan-23 18:11:58 — root — ERROR — <module>:808 — Reading data file test.info
23:42:12  Found 1296 entries.
23:42:12  Using user-specified filename prefix "/home/root/workspace/test-build"
23:42:12  Writing .css and .png files.
23:42:12  Generating output.
23:42:12  genhtml: ERROR: cannot create directory build/test!

Checked the permission for the path and even after providing 777 permission on the folders it is failing with same error.
In the html output directory below files are created when command failed  : 
amber.png  emerald.png  gcov.css  glass.png  ruby.png  snow.png  updown.png

lcov version : 1.16
lcov --summary test.info , this command is working fine with the info file.

**Comment by henry2cox** [comment]
Which version of lcov are you using?  (TOT or some released version?)
If not TOT:  could you try the most recent github code, and see if it tells you anything different?
If you DO still see an error:   please run again with flags "-v -v" (turn up verbosity) and "--debug" (turn on some debug dumping).
The other question is whether this fails the same way if you write output to somewhere else (say, to your user home directory).

**Comment by rachit16** [comment]
@henry2cox  I am using released version 1.16. yes it fails in same way if i write the output in any directory (home or tmp). The size of my info file is around 4.6 GB. 

Tried passing -v -v with genhtml , it is giving me version details and for debug i am getting unknown option.

genhtml: Unknown option: debug
Use genhtml --help to get usage information

**Comment by henry2cox** [comment]
Right...those options are for a newer version of code than you are using.
Interesting that things work the way you expect when branch coverage is disabled.  That ought not to affect destinations.

I think that the best thing to do is to try a newer version of code - and then we can debug from there.

**Comment by rachit16** [comment]
genhtml --legend -p /home/root/workspace/test-build test.info -o /tmp/test --ignore-errors source -v  --dark-mode --keep-going --tempdir /tmp/test_tmp_dir --preserve --branch-coverage --function-coverage test.info --profile

ran this command with newer version of lcov by cloning from github.
It is running since last 8 hours , below are the console logs so far

Reading data file test.info
Found 1296 entries.
Using user-specified filename prefix "/home/aviuser/workspace/eng-ci-build"
Writing .css and .png files.
genhtml: ERROR: cannot open gcov.css for writing!

Lcov version after cloning from github: 
lcov: LCOV version v1.16-16-g038c2ca

**Comment by henry2cox** [comment]
I guess that's progress :-)  At least still running rather than dying right away.
8 hours is long, though.

If you have a multicore server and a fair bit of memory, you could try adding "--parallel ..." to the command line.
That will cause genhtml to work across multiple cores - but will require more memory to do so.
There is a second flag "--memory size_in_Mb" that tells the tool to throttle parallel execution so as to not exceed some limit.  (This is necessary in some LSF configurations - which might otherwise just kill you without recourse.)
We see near linear speedup, for up to about 20 cores - then some tailoff after that (all else being equal).  Network, filesystem, and other limitations can make things go slower than we might like.

**Comment by rachit16** [comment]
I added --keep-going arguments , is this can be the reason it is not dying ?

**Comment by henry2cox** [comment]
Forgot to mention that you are unlikely to be happy with the default branch coverage report, especially if you are using C++.
You might want to look at the discussion in issue 101 (now closed) - or simply add "--filter branch" to your lcov or genhtml command line.
Note that filtering _does_ require that your source code be available/visible.  That might require path hacking (see the "--substitute" option).

**Comment by rachit16** [comment]
Unfortunately Command exited after running for more than 10 hours with below console log 

Found 1296 entries.
Using user-specified filename prefix "/home/aviuser/workspace/eng-ci-build"
Writing .css and .png files.
genhtml: ERROR: cannot open gcov.css for writing!

I checked the output directory /tmp/test , Nothing is written in the directory.

**Comment by henry2cox** [comment]
Changes mentioned in '3.', above are pushed.

**Comment by rachit16** [comment]
@henry2cox i tried with 5 the entries , it went fine. Is there any way where we can check on which specific SF part we are having any issue ? I have 600+ test cases result which i combined to get single info file using lcov -a option.  when lcov -a command ended it displays line , function and branch coverage metrics but when i am trying to generate html from the created info file facing this issue.

May be if we can add any command line option in genhtml or lcov to validate info file in future releases it will be helpful to debug such issue i believe :) 

For this scenario is there any way to validate info file and get the corrupted entry which is causing this issue ?

**Comment by rachit16** [comment]
> Changes mentioned in '3.', above are pushed.

Let me run the genhtml again with the latest changes. I ran it again,  it will take around 10-12 hours to complete. genhtml doen't have --debug option , i triggered with -v -v option.

**Comment by rachit16** [comment]
@henry2cox Ran the same command with latest changes below are the logs before it exited  

genhtml: Use of uninitialized value $which in string eq at /usr/local/bin/genhtml line 1867.
genhtml: Use of uninitialized value $which in string eq at /usr/local/bin/genhtml line 1870.
genhtml: Use of uninitialized value $which in concatenation (.) or string at /usr/local/bin/genhtml line 1870.
genhtml: unknown key  - should be 'base' or 'current' at /usr/local/bin/genhtml line 1870.

I can see exit status of the genhtml command is 25.

In the output directory it has created the build directory where it was failing earlier but index.html is absent in the output directory.

ls /root/reports/test/html-test
amber.png  build  emerald.png  gcov.css  glass.png  infra  ptbf  ruby.png snow.png  tpd  updown.png

**Comment by henry2cox** [comment]
The exit status you see is from a 'die' call that happens with the method decides that it has been called with an invalid set of parameters ("undef", in this case).
Unfortunately, I don't immediately see how that is happening...a testcase would be useful.
Failing that, though:  can you re-run your example with some more Perl debugging turned on?

  $ perl -Mdiagnostics=-t,-w $LCOV_HOME/bin/genhtml ...  <- your command line

The result is likely to be a bit more verbose - but I hope that we will get a full stack trace at the point of the error, so I can see what went wrong.

**Comment by rachit16** [comment]
> The exit status you see is from a 'die' call that happens with the method decides that it has been called with an invalid set of parameters ("undef", in this case). Unfortunately, I don't immediately see how that is happening...a testcase would be useful. Failing that, though: can you re-run your example with some more Perl debugging turned on?
> 
> $ perl -Mdiagnostics=-t,-w $LCOV_HOME/bin/genhtml ... <- your command line
> 
> The result is likely to be a bit more verbose - but I hope that we will get a full stack trace at the point of the error, so I can see what went wrong.

Sure i have triggered it again with perl -Mdiagnostics=-t,-w <Mycommand> , let me share the result once it completes.

**Comment by henry2cox** [comment]
Actually...I think I figured it out ...
This die is actually happening as the code is trying to report another issue (inconsistent branch data)
I'll push a fixed message shortly ... but then we are going to need to figure out why the data is wrong/inconsistent, and what genhtml should do about it.
(Don't kill your other experiment with perl diagnostics, though - my analysis could be wrong)

**Comment by henry2cox** [comment]
One more set of questions.  I think I know part of what is happening
Can you describe how you are using lcov and some information about your project?

- what languages?  (C, C++, both, other languages as well
- toolchain and version (gcc, llvm, other)
- approximately how many files?  How many unique directories?  How many LOC?
- how many testcases?
- what is your 'lcov --capture' command?  Your 'lcov aggregate' command?

My suspicion is that your coverage.info file is huge because you capture each test result as a unique testcase.  As a result, the .info contains line, branch, and function entries for every file, for every testcase.  Then genhtml is merging the testcase data to present a unified view (...and that is taking a long time).   Branch data is inconsistent due to parameterization of some temples in some files for some testcases.

If the above guess is true - then:  interesting.

- This is a supported use model - and so it should work.  However:  this _is_ different than any of the large projects that I am currently aware of ("large" == more than 500K executable lines).  You are likely running into some capacity issues that need to be fixed.
- On the other hand, if you are not interested in differences between particular testcases - then an alternate use model may be more appropriate, and will execute much faster.


Henry

**Comment by rachit16** [comment]
> 

@henry2cox  Below are my comments on the questions asked 
what languages? (C, C++, both, other languages as well) - C and C++ both we are using
toolchain and version (gcc, llvm, other) gcc 9.4.0
approximately how many files? How many unique directories? How many LOC? - Files are more than 20k i combine artifacts using gcov-tool and create 600+ info files.
how many testcases? 600+
what is your 'lcov --capture' command? Your 'lcov aggregate' command? 
lcov -c -d infra -t test1 --no-external -b /home/root/workspace/test --rc lcov_branch_coverage=1 -o test1.info
For combining info files
lcov --rc lcov_branch_coverage=1 -a test1 -a test2 -o final.info

Size of final.info is around 4.8GB usually.

then i use genhtml command to generate the html out of it. It used to work always for me till last week, then it suddenly started failing with the error which i mentioned in this issue.

Job usually runs for 8-9 hours to generate the report. Is there any way i can improve the performance for this scenario.
System on which i run these lcov and genhtml command is having 48Gb RAM and 120 GB and 16 cores CPU. Resource wise i dont see any issue on my machine during whole execution process.

Thank you

**Comment by henry2cox** [comment]
It appears that you don't care about particular testcases - only their aggregate.
That is:  you really just want to merge the coverage data from all the testcases into a single report such that coverpoints are considered to be covered if they are exercised by any testcase.  You are specifically NOT trying to figure out which points are covered by what tests (e.g., for some knapsack/testcase reduction procedure).

In that case:  you can simply remove the testname parameter from the capture command:
  $ lcov -c -d infra --no-external -b /home/root/workspace/test --branch-coverage -o test1.info
(This will result in the special testname "" - empty string.)

Then you can merge all those .info files as you have been doing:
  $ lcov --branch-coverage -o final.info -a ....   <- you can make this go faster if you use a glob pattern for the "-a" argument
All these files have the same testname - so get merged into one report.

Both these steps are likely to go faster if you turn on a bit of parallelism:
  $ lcov --parallel -c -d infra --no-external -b /home/root/workspace/test --branch-coverage -o test1.info
  $ lcov --parallel --branch-coverage -o final.info -a 'test*.info'
genhtml will also go faster, with parallelism.
(See the man pages for a discussion of the --parallel option, as well as for throttling controls.)

If you also use the "--profile" option, then the various commands will write out data file that can be used to investigate performance - i.e., to see where time is going, and hopefully identify some bottlenecks that can be fixed.  See the man pages.

Using some of the above + a couple of other tricks - we made one of our jobs go well over 100X faster than it had been (more than 10000%) :-)  Such a performance improvement changes the use model - from (say) once per week/project wide to commit criteria.
YMMV.

**Comment by rachit16** [comment]
@henry2cox if i will remove -t from lcov command i will not be able to get the which all tests are hitting particular file. In the report generated i generally get this information as shown in attached image. Yes i am intrested in combined report for all test cases but need test names as well like attached one to judge which job is hitting which file.

<img width="1177" alt="image" src="https://user-images.githubusercontent.com/18699253/210800572-6894e73a-5dec-4c3c-a6de-9c7b36a10552.png">

**Comment by henry2cox** [comment]
For sorting/deciding which tests to drop, you might want to look at the lcov '--prune-tests' and '--map-functions' options.
Not quite the same application as what you may be looking for - but possibly useful.

**Comment by rachit16** [comment]
> > The exit status you see is from a 'die' call that happens with the method decides that it has been called with an invalid set of parameters ("undef", in this case). Unfortunately, I don't immediately see how that is happening...a testcase would be useful. Failing that, though: can you re-run your example with some more Perl debugging turned on?
> > $ perl -Mdiagnostics=-t,-w $LCOV_HOME/bin/genhtml ... <- your command line
> > The result is likely to be a bit more verbose - but I hope that we will get a full stack trace at the point of the error, so I can see what went wrong.
> 
> Sure i have triggered it again with perl -Mdiagnostics=-t,-w , let me share the result once it completes.

@henry2cox Seems like this has not printed anything new. It printed same log which i shared earlier.

genhtml: Use of uninitialized value $which in string eq at /usr/local/bin/genhtml line 1867.
genhtml: Use of uninitialized value $which in string eq at /usr/local/bin/genhtml line 1870.
genhtml: Use of uninitialized value $which in concatenation (.) or string at /usr/local/bin/genhtml line 1870.
genhtml: unknown key - should be 'base' or 'current' at /usr/local/bin/genhtml line 1870.

**Comment by henry2cox** [comment]
Damn.
Now that is rather disappointing.

Can you grab the update I pushed a few hours back?  It should turn that ‘die’ into an ignorable error – and your command line already contains “--keep-going” – so the result should be to emit a message and continue.

Meantime, I will do a bit more investigation to see how to get a stacktrace.

Henry

**Comment by rachit16** [comment]
> Damn. Now that is rather disappointing. Can you grab the update I pushed a few hours back? It should turn that ‘die’ into an ignorable error – and your command line already contains “--keep-going” – so the result should be to emit a message and continue. Meantime, I will do a bit more investigation to see how to get a stacktrace. Henry

@henry2cox yes already started few minutes back with latest changes. Will share result as soon as it completes.

**Comment by rachit16** [comment]
I saw one error which might help us in finding invalid entry in info file.

Processing file build/test/ring3/include/ring3.h
genhtml: Error: 'current' line 349 (349) branch 0 contains 8 elements but merge data contains 4.

**Comment by henry2cox** [comment]
Yeah … that will help.
Can you look in your .info file to find all of the “SF:build/test/ring3/include/ring3.h” lines, then all of the “BRDA:349,…” lines in that file record
(these are the branches in question).  Each “TN:…” record will have one of the SF: records if that testcase used the file (so we expect to see up to about 600 such entries).

The more basic question is what the code on line 349 looks like – or really, what the function/method containing that line looks like.
My guess is “C++ template function”, possibly with parameters that are themselves templates – and the conditional on line 349 uses at least one of those template parameters…
If this is the case…then we are getting wrapped up in some compler-generated exception handling code.

Depending on what the code looks like:  I may or may not have suggestions how to work around or repair the issue.

A related question:  does the genhtml process complete now?  Or does it get stuck somewhere else?
Are there any other interesting looking error/warning messages to be seen?

Thanks

Henry

**Comment by rachit16** [comment]
> 

@henry2cox It is still running. These are last few lines on console, but seems genhtml is stucked here since last 3-4 hours.

Filter suppressions 'branch':
    10829189 instances
    61816569 coverpoints
genhtml: Use of uninitialized value in concatenation (.) or string at /usr/local/bin/genhtml line 5470.
chdir /root/info (from )

But i can see index.html files are being generated in output directory now.

Code is still running.

Line Number 349 of the code looks like below 

349: if (n == 0)
350:               <Some Code>;

Content of info file 
BRDA:349,0,0,7235390000
BRDA:349,0,1,184907
BRDA:349,0,2,1847990000
BRDA:349,0,3,80433
BRDA:349,0,4,44919900
BRDA:349,0,5,0
BRDA:349,0,6,-
BRDA:349,0,7,-

**Comment by rachit16** [comment]
> After that “chdir ..” message, the only thing that should be happening is to emit the profile data (if you passed the “--profile” flag) and to check your coverage criteria (if you passed the ‘’--criteria-script’ flag) – so the job should be finished or very nearly finished. Not sure why it would hang for a long time. Is the perl process still running? Or hung/stopped somewhere? (you might be able to simply kill it). You should also be able to look at the HTML in your browser, to see if everything looks OK or if there are obvious errors. There should be more than one .info entry for that file and branch – I think that we are interested to see some of the others. There will be at least one that has 4 branch elements – and possibly others that have some number other than 8 or 4. That we see multiple branch elements on the line containing a single ‘if’ suggests that the compiler has located a bunch of exception handers there. What does the rest of the function look like? As we see…the branch filter sees that there is a conditional here – so doesn’t try to filter anything. As a result: we see 8 branches despite that the source code has only 2 (taken/not taken).

@henry2cox  Is the perl process still running?  Or hung/stopped somewhere?  (you might be able to simply kill it). - Process completed just now. I have not passed --profile option so not sure why it was hunged up for longer time.

Report looks good to me. It just got completed but do not printed anything after last console log which i shared earlier. 

I checked all BRDA:349 entries for the file for which error was showing in console log , They all are having 4 fields when i splitted it with comma.

**Comment by henry2cox** [comment]
Sounds good.

Note that the BRDA records contain 4 values.  A set of adjacent values with the same line and 'block' ID refers to the same branch expression.  The sample you sent reflects 8 branches on line 349 - but the error message says that there is at least one other expression containing 4 branches from some other testcase region in the .info file.
I don't think we know quite why that is happening yet - but I don't think I can figure it out without seeing the function in question and the associated coverage data.

In any event, I think that we can claim that the original issue (failed mkdir) is either fixed or did not happen in your recent jobs.
It also seems that the 'die' issue related to branchcov merging is either fixed or worked around.

Given that, I will push a few more functional changes that have been pending locally for a while.
This also contains a slightly improved 'branchcov inconsistency' message.
Please update and try your testcase again - and let me know what happens. 
If it still works:  then I think we can declare success, and close this issue.

Thanks
Henry

**Comment by rachit16** [comment]
@henry2cox Thanks for your analysis on this issue and helping me on resolving this issue. I will retry with new test data set and confirm you. Will try the optimisations you shared in this tread as well.

**Comment by henry2cox** [comment]
Checking if there are any lingering problems - or if this issue can be closed.

**Comment by rachit16** [comment]
@henry2cox I will be running the report on new test data tomorrow then only i can confirm on this ? Is it ok to keep this issue open for 2 more days ?

**Comment by henry2cox** [comment]
No worries.
Tomorrow is Saturday, though – so perhaps you should take the day off, and not worry much about code coverage. 😊

(I’m just trying to make sure that issues are resolved and don’t linger for long periods of time.
Would prefer to fix quickly so others don’t hit the same problem.)

**Comment by rachit16** [comment]
@henry2cox I can confirm , issue is resolved now. Thanks for your help !!

**Comment by henry2cox** [comment]
Good to hear that it is fixed/works.
I'm curious if you fixed your performance issues as well.
From what I see, above - you should be able to get this to run in under an hour - with luck and a nontrivial server box:  a bit under half an hour.  (Yeah - this issue report is not the place to discuss performance improvements...perhaps we should open the "community" pages for discussions.)
Henry
